lib/Frontend/FrontendInputsAndOutputs.cpp

```diff
@@ -238,3 +256,90 @@ void FrontendInputsAndOutputs::addPrimaryInputFile(StringRef file,
 }
+
+// Outputs
+
+unsigned FrontendInputsAndOutputs::countOfInputsProducingMainOutputs() const {
+  return isSingleThreadedWMO()
+             ? 1
+             : hasPrimaryInputs() ? primaryInputCount() : inputCount();
+}
+
+const InputFile &FrontendInputsAndOutputs::firstInputProducingOutput() const {
+  return isSingleThreadedWMO()
+             ? firstInput()
+             : hasPrimaryInputs() ? firstPrimaryInput() : firstInput();
+}
+
+const InputFile &FrontendInputsAndOutputs::lastInputProducingOutput() const {
+  return isSingleThreadedWMO()
+             ? firstInput()
+             : hasPrimaryInputs() ? lastPrimaryInput() : lastInput();
+}
+
+void FrontendInputsAndOutputs::forEachInputProducingAMainOutputFile(
+    llvm::function_ref<void(const InputFile &)> fn) const {
+  isSingleThreadedWMO()
+      ? fn(firstInput())
+      : hasPrimaryInputs() ? forEachPrimaryInput(fn) : forEachInput(fn);
+}
+
+void FrontendInputsAndOutputs::setMainOutputs(
+    ArrayRef<std::string> outputFiles) {
+  assert(countOfInputsProducingMainOutputs() == outputFiles.size());
+  if (hasPrimaryInputs()) {
+    unsigned i = 0;
+    for (auto index : indices(AllInputs)) {
+      InputFile &f = AllInputs[index];
+      if (f.isPrimary())
+        f.setOutputFilename(outputFiles[i]);
+    }
+  } else if (isSingleThreadedWMO()) {
+    AllInputs[0].setOutputFilename(outputFiles[0]);
+  } else {
+    for (auto i : indices(AllInputs))
+      AllInputs[i].setOutputFilename(outputFiles[i]);
+  }
+}
+
+std::vector<std::string> FrontendInputsAndOutputs::copyOutputFilenames() const {
+  std::vector<std::string> outputs;
+  forEachInputProducingAMainOutputFile([&](const InputFile &input) -> void {
+    outputs.push_back(input.outputFilename());
+  });
+  return outputs;
+}
+
+void FrontendInputsAndOutputs::forEachOutputFilename(
+    llvm::function_ref<void(const std::string)> fn) const {
+  forEachInputProducingAMainOutputFile(
+      [&](const InputFile &input) -> void { fn(input.outputFilename()); });
+}
+
+StringRef FrontendInputsAndOutputs::getSingleOutputFilename() const {
+  assertMustNotBeMoreThanOnePrimaryInputUnlessBatchModeChecksHaveBeenBypassed();
+  return hasInputs() ? StringRef(lastInputProducingOutput().outputFilename())
+                     : StringRef();
+}
+
+bool FrontendInputsAndOutputs::isOutputFilenameStdout() const {
+  return getSingleOutputFilename() == "-";
+}
+
+bool FrontendInputsAndOutputs::isOutputFileDirectory() const {
+  return hasNamedOutputFile() &&
+         llvm::sys::fs::is_directory(getSingleOutputFilename());
+}
+
+bool FrontendInputsAndOutputs::hasNamedOutputFile() const {
+  return hasInputs() && !isOutputFilenameStdout();
+}
+
+// Supplementary outputs
+
+void FrontendInputsAndOutputs::forEachInputProducingSupplementaryOutput(
+    llvm::function_ref<void(const InputFile &)> fn) const {
+  isWholeModule()
+      ? fn(firstInput())
+      : hasPrimaryInputs() ? forEachPrimaryInput(fn) : forEachInput(fn);
```

Comment:
I don't think there are any situations where both `isWholeModule()` and `hasPrimaryInputs()` are false.

Reply:
Nice catch! Redone:
    if (hasPrimaryInputs())
      forEachPrimaryInput(fn);
    else
      fn(firstInput());